File: Leetcode/19-Remove-Nth-Node-From-End-of-List/RemoveNthNodeFromEndOfList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if (!head || n<0) {
        return null
    }
    let dummy = new ListNode(0)
    dummy.next = head
    var prev = dummy
    var target = head
    var tail = head
    var i
    for (i=0; i<n && tail; i++) {
        tail = tail.next
    }
    while (tail) {
        tail = tail.next
        target = target.next
        prev = prev.next
    }
    prev.next = target.next
    return dummy.next
};

File: Leetcode/19-Remove-Nth-Node-From-End-of-List/RemoveNthNodeFromEndOfList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil || n < 0 {
		return nil
	}
	dummy := &ListNode{Val: 0, Next: head}
	prev := dummy
	target := head
	tail := head
	i := 0
	for i < n && tail != nil {
		tail = tail.Next
		i++
	}
	for tail != nil {
		tail = tail.Next
		target = target.Next
		prev = prev.Next
	}
	prev.Next = target.Next
	return dummy.Next
}
